subsys/usb/usb_device.c

```diff
@@ -1019,3 +1047,181 @@ int usb_ep_read_continue(u8_t ep)
+
+	/* restore endpoint callback */
+	usb_dc_ep_set_callback(trans->ep, usb_dev.config->endpoint[epid].ep_cb);
+
+	trans->status = status;
+	k_sem_give(&trans->sem);
+
+	if (trans->cb)
+		trans->cb(trans->ep, trans->status, trans->size);
+}
+
+static void usb_handle_transfer(u8_t ep, enum usb_dc_ep_cb_status_code status)
+{
+	struct usb_transfer_data *trans;
+	int ret;
+
+	trans = usb_ep_get_transfer(ep);
+	if (!trans) {
+		return;
+	}
+
+	if (trans->status != -EBUSY) {
+		return;
+	}
+
+	if (status == USB_DC_EP_DATA_IN) { /* write complete */
+		int bytes;
+
+		if (!trans->bsize) { /* transfer complete */
+			if (trans->zero)
+				usb_dc_ep_write(ep, NULL, 0, NULL);
+			usb_transfer_done(trans, 0);
+			return;
+		}
+
+		ret = usb_dc_ep_write(ep, trans->buffer, trans->bsize, &bytes);
+		if (ret) {
+			/* transfer error */
+			usb_transfer_done(trans, -EINVAL);
+			return;
+		}
+
+		trans->buffer += bytes;
+		trans->bsize -= bytes;
+		trans->size += bytes;
+	} else if (status == USB_DC_EP_DATA_OUT) {
+		int bytes;
+
+		ret = usb_dc_ep_read(ep, trans->buffer, trans->size, &bytes);
+		if (ret) {
+			/* transfer error */
+			usb_transfer_done(trans, -EINVAL);
+			return;
+		}
+
+		trans->buffer += bytes;
+		trans->bsize -= bytes;
+		trans->size += bytes;
+
+		/* ZLP, short-pkt or buffer full */
+		if (!bytes || (bytes % usb_dc_ep_mps(ep)) || !trans->bsize) {
+			/* transfer complete */
+			usb_transfer_done(trans, 0);
+			return;
+		}
+	}
+}
+
+int usb_transfer(u8_t ep, u8_t *data, size_t dlen, unsigned int flags,
+		 usb_transfer_callback cb)
+{
+	struct usb_transfer_data *trans = usb_ep_get_transfer(ep);
+	unsigned int key;
+	int ret;
+
+	SYS_LOG_DBG("ep=%02x, data=%p, dlen=%d\n", ep, data, dlen);
+
+	if (!trans) {
+		return -EINVAL;
+	}
+
+	if (k_sem_take(&trans->sem, K_NO_WAIT)) {
+		return -EBUSY;
+	}
+
+	/* Configure transfer */
+	trans->ep = ep;
+	trans->buffer = data;
+	trans->bsize = dlen;
+	trans->size = 0;
+	trans->cb = cb;
+	trans->status = -EBUSY;
+
+	/* we need to temporaly change the ep callback */
+	usb_dc_ep_set_callback(ep, usb_handle_transfer);
+
+	key = irq_lock();
+	if (flags & USB_TRANS_WRITE) {
+		usb_handle_transfer(ep, USB_DC_EP_DATA_IN);
+	} else {
+		int bytes;
+
+		/* Already have data to read ? */
+		ret = usb_dc_ep_read(ep, NULL, 0, &bytes);
+		if (!ret && bytes) {
+			usb_handle_transfer(ep, USB_DC_EP_DATA_OUT);
+		}
+	}
+	irq_unlock(key);
+
+	if (!(dlen % usb_dc_ep_mps(ep))) {
+		/* ZLP needed ? */
+		trans->zero = true;
+	}
+
+	return 0;
+}
+
+int usb_transfer_sync(u8_t ep, u8_t *data, size_t dlen, unsigned int flags)
+{
+	struct usb_transfer_data *trans = usb_ep_get_transfer(ep);
+	int ret;
+
+	ret = usb_transfer(ep, data, dlen, flags, NULL);
+	if (ret) {
+		return ret;
+	}
+
+	/* wait for transfer completion */
+	if (k_sem_take(&trans->sem, K_FOREVER)) {
+		trans->status = -ETIMEDOUT;
+	}
+
+	if (!trans->status) {
+		ret = trans->size;
+	} else {
+		ret = trans->status;
+	}
+
+	/* restaure ep callback */
```

Comment:
"restaure" is the French word for restore? :)

Reply:
oops ;-)